lib/acyclicGraph.mli

```diff
@@ -0,0 +1,79 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** Graphs representing strict orders *)
+
+type constraint_type = Lt | Le | Eq
+
+module type Point = sig
+  type t
+
+  module Set : CSig.SetS with type elt = t
+  module Map : CMap.ExtS with type key = t and module Set := Set
+
+  module Constraint : CSet.S with type elt = (t * constraint_type * t)
+
+  val equal : t -> t -> bool
+  val compare : t -> t -> int
+
+  type explanation = (constraint_type * t) list
+  val error_inconsistency : constraint_type -> t -> t -> explanation lazy_t option -> 'a
+
+  val pr : t -> Pp.t
+end
+
+module Make (Point:Point) : sig
+
+  type t
+
+  val empty : t
+
+  val check_invariants : required_canonical:(Point.t -> bool) -> t -> unit
+
+  exception AlreadyDeclared
+  val add : ?rank:int -> Point.t -> t -> t
+  (** Use a large [rank] to keep the node canonical *)
+
+  exception Undeclared of Point.t
+  val check_declared : t -> Point.Set.t -> unit
+
+  type 'a check_function = t -> 'a -> 'a -> bool
+
+  val check_eq : Point.t check_function
+  val check_leq : Point.t check_function
+  val check_lt : Point.t check_function
+
+  val enforce_eq : Point.t -> Point.t -> t -> t
+  val enforce_leq : Point.t -> Point.t -> t -> t
+  val enforce_lt : Point.t -> Point.t -> t -> t
```

Comment:
Do these functions assume that the `Point.t` arguments have already been added to the graph? (using `add` I suppose)
If yes then it might be worthwhile to make that explicit in a comment.

Reply:
@SkySkimmer?

Reply:
comment added